engine/src/main/java/org/hibernate/validator/spi/cascading/ValueExtractor.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.spi.cascading;
+
+public interface ValueExtractor<T> {
+
+	void extractValues(T originalValue, ValueReceiver receiver);
+
+	interface ValueReceiver {
+
+		void objectValue(Object object);
+
+		void listValue(int i, Object object);
```

Comment:
OK, additional comment (but probably useless if it goes away): as you inversed the order of key and value in `mapValue`, I think you should do the same for `listValue` and have the index as second parameter. It would be more consistent.